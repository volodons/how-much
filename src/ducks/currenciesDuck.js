import { createSlice } from '@reduxjs/toolkit';

import ALL_CURRENCIES from '../const/currencies';

const initialState = {
    allCurrencies: [...ALL_CURRENCIES],
    featuredCurrencies: [],
};

const currenciesSlice = createSlice({
    name: 'Currencies',
    initialState,
    reducers: {
        ADD_TO_FEATURED_CURRENCIES: (state, action) => {
            state.featuredCurrencies = [
                ...state.featuredCurrencies,
                action.payload,
            ];
            state.allCurrencies = state.allCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
        },
        REMOVE_FROM_FEATURED_CURRENCIES: (state, action) => {
            state.featuredCurrencies = state.featuredCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
            state.allCurrencies = [...state.allCurrencies, action.payload];
        },
    },
});

const { actions, reducer } = currenciesSlice;

export const currenciesActions = actions;
export const currenciesReducer = reducer;
