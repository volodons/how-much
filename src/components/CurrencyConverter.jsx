import { useDispatch, useSelector } from "react-redux";
import { exchangeRatesActions } from "../ducks/exchangeRatesDuck";
import CURRENCIES from "../const/currencies";
import Flag from "react-world-flags";
import CurrencyExchangeIcon from "@mui/icons-material/CurrencyExchange";
import Autocomplete from "@mui/material/Autocomplete";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import IconButton from "@mui/material/IconButton";
import Input from "@mui/material/Input";
import { styled } from "@mui/material/styles";
import TextField from "@mui/material/TextField";
import Typography from "@mui/material/Typography";

const StyledTypographyHeading = styled(Typography)`
    margin-bottom: 2rem;
    text-align: center;
    color: #f5f7ff;
    text-decoration: underline;
    font-family: Lobster, sans-serif;
    text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    &:hover {
        text-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
    }
    @media (max-width: 600px) {
        font-size: 2.5rem;
    }
`;

const StyledTypographyButton = styled(Typography)`
    font-size: 1.25rem;
    font-weight: bold;
`;

const StyledBox = styled(Box)`
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 1rem;
    padding: 1rem;
    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    border-radius: 10px;
    background: linear-gradient(to right, #0390fc, #562fff);
    transition: box-shadow 0.3s ease-in-out;
    &:hover {
        box-shadow: 4px 4px 8px rgba(0, 0, 0, 0.5);
        transform: scale(1.02);
    }
    @media (max-width: 900px) {
        width: 75%;
    }
    @media (max-width: 600px) {
        width: auto;
        max-width: 100%;
    }
`;

const StyledOuterGridContainer = styled(Grid)`
    display: flex;
    flex-direction: column;
    align-items: center;
    row-gap: 1rem;
`;

const StyledInnerGridContainer = styled(Grid)`
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
    column-gap: 5rem;
    row-gap: 1rem;
`;

const StyledGridItem = styled(Grid)`
    width: 100%;
`;

const StyledInput = styled(Input)`
    width: 100%;
    text-align: center;
    border-color: #f5f7ff;
    & .MuiInputBase-input {
        text-align: center;
        color: #ffffff;
    }
`;

const StyledTextFiled = styled(TextField)`
    & .MuiOutlinedInput-root {
        &:hover fieldset {
            border-color: #f5f7ff;
        }
    }
    & .MuiAutocomplete-inputRoot {
        color: #f5f7ff;
    }
    & .MuiButtonBase-root {
        color: #f5f7ff;
    }
    & .MuiInputLabel-root {
        color: #f5f7ff;
        &:active {
            color: #f5f7ff;
        }
        &:focus {
            color: #f5f7ff;
        }
    }
    & .MuiOutlinedInput-notchedOutline {
        border-width: 2px;
        border-color: #f5f7ff;
        border-radius: 10px;
        &:active {
            border-color: #f5f7ff;
        }
        &:focus {
            border-color: #f5f7ff;
        }
    }
`;

const StyledIconButton = styled(IconButton)`
    gap: 0.5rem;
    color: #f5f7ff;
    border: 2px solid #f5f7ff;
    border-radius: 10px;
    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    &:hover {
        box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
    }
`;

const StyledFlag = styled(Flag)`
    width: 2rem;
    height: 2rem;
    margin-right: 1rem;
`;

const StyledCurrencyExchangeIcon = styled(CurrencyExchangeIcon)`
    font-size: 2rem;
    color: #f5f7ff;
`;

const CurrencyConverter = () => {
    const dispatch = useDispatch();
    const {
        amountBeforeConversion,
        amountAfterConversion,
        baseCurrency,
        targetCurrency,
    } = useSelector((state) => state.exchangeRates);

    const handleChangeAmountBeforeConversion = (amount) => {
        dispatch({
            type: exchangeRatesActions.SET_AMOUNT_BEFORE_CONVERSION.type,
            payload: parseFloat(amount),
        });
    };

    const handleChangeBaseCurrency = (currency) => {
        dispatch({
            type: exchangeRatesActions.SET_BASE_CURRENCY.type,
            payload: currency,
        });
    };

    const handleChangeTargetCurrency = (currency) => {
        dispatch({
            type: exchangeRatesActions.SET_TARGET_CURRENCY.type,
            payload: currency,
        });
    };

    const handleConvertCurrency = () => {
        dispatch({ type: exchangeRatesActions.FETCH_EXCHANGE_RATES.type });
    };

    return (
        <StyledBox>
            <StyledTypographyHeading variant="h2">
                Currency Converter
            </StyledTypographyHeading>
            <StyledOuterGridContainer container>
                <StyledInnerGridContainer container>
                    <Grid container>
                        <StyledGridItem item>
                            <StyledInput
                                placeholder="Type in number"
                                type="number"
                                required
                                value={amountBeforeConversion || ""}
                                onChange={(event) =>
                                    handleChangeAmountBeforeConversion(
                                        event.target.value
                                    )
                                }
                            />
                        </StyledGridItem>
                    </Grid>
                    <Grid item>
                        <Autocomplete
                            value={baseCurrency || ""}
                            options={CURRENCIES}
                            getOptionLabel={(option) =>
                                option.currency || baseCurrency
                            }
                            renderInput={(params) => (
                                <StyledTextFiled
                                    {...params}
                                    label="Choose Currency To Convert"
                                    required
                                />
                            )}
                            renderOption={(props, option) => (
                                <Box component="li" {...props}>
                                    <StyledFlag code={option.country} />
                                    {option.currency}
                                </Box>
                            )}
                            onChange={(event, value) =>
                                handleChangeBaseCurrency(value.currency)
                            }
                        />
                    </Grid>
                    <Grid item>
                        <Autocomplete
                            value={targetCurrency || ""}
                            options={CURRENCIES}
                            getOptionLabel={(option) =>
                                option.currency || targetCurrency
                            }
                            renderInput={(params) => (
                                <StyledTextFiled
                                    {...params}
                                    label="Choose Currency To Convert Into"
                                    required
                                />
                            )}
                            renderOption={(props, option) => (
                                <Box component="li" {...props}>
                                    <StyledFlag code={option.country} />
                                    {option.currency}
                                </Box>
                            )}
                            onChange={(event, value) =>
                                handleChangeTargetCurrency(value.currency)
                            }
                        />
                    </Grid>
                </StyledInnerGridContainer>
                <StyledIconButton onClick={handleConvertCurrency}>
                    <StyledCurrencyExchangeIcon />
                    <StyledTypographyButton variant="button">
                        Convert
                    </StyledTypographyButton>
                </StyledIconButton>
                <StyledInnerGridContainer container>
                    <StyledGridItem item>
                        <StyledInput
                            value={amountAfterConversion || ""}
                            placeholder="Result"
                            type="number"
                            readOnly
                        />
                    </StyledGridItem>
                </StyledInnerGridContainer>
            </StyledOuterGridContainer>
        </StyledBox>
    );
};

export default CurrencyConverter;
