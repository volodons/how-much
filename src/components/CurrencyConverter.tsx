import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Autocomplete from '@mui/material/Autocomplete';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';

import { RootState } from '../redux';
import { converterActions } from '../redux/ducks/converterDuck';
import { currenciesActions } from '../redux/ducks/currenciesDuck';
import {
    StyledCircularProgressBox,
    StyledCircularProgress,
    StyledTypographyHeading,
    StyledTypographyButton,
    StyledCurrencyConverterBox,
    StyledOuterGridContainer,
    StyledInnerGridContainer,
    StyledGridItem,
    StyledInput,
    StyledTextField,
    StyledIconButton,
    StyledCurrencyExchangeIcon,
} from '../styled/styledCurrencyConverter';

interface Currency {
    id: number;
    code: string;
    exchangeRate: number;
    featured: boolean;
}
interface Values {
    baseCurrency: Currency | string;
    baseCurrencyAmount: number;
    targetCurrency: Currency | string;
    targetCurrencyAmount: number;
}

const CurrencyConverter: React.FC = () => {
    const dispatch = useDispatch();
    const { currencies, targetCurrencyAmount } = useSelector(
        (state: RootState) => state.converter
    );
    const { baseCurrency } = useSelector(
        (state: RootState) => state.currencies
    );

    const handleConvertCurrency = (values: Values) => {
        dispatch({
            type: converterActions.FETCH_EXCHANGE_RATE.type,
            payload: {
                baseCurrency: (values.baseCurrency as Currency).code,
                targetCurrency: (values.targetCurrency as Currency).code,
            },
        });
    };

    useEffect(() => {
        dispatch({ type: converterActions.FETCH_CURRENCIES.type });
    }, []);

    useEffect(() => {
        dispatch({ type: currenciesActions.FETCH_BASE_CURRENCY.type });
    }, []);

    const initialValues: Values = {
        baseCurrency: baseCurrency || '',
        baseCurrencyAmount: 0,
        targetCurrency: '',
        targetCurrencyAmount: targetCurrencyAmount || 0,
    };

    return (
        <Box data-testid="currency-converter-component">
            {!currencies || !baseCurrency ? (
                <StyledCircularProgressBox>
                    <StyledCircularProgress size="5rem" />
                </StyledCircularProgressBox>
            ) : (
                <StyledCurrencyConverterBox>
                    <StyledTypographyHeading variant="h2">
                        Currency Converter
                    </StyledTypographyHeading>
                    <Formik
                        initialValues={initialValues}
                        onSubmit={(values) => {
                            dispatch({
                                type: converterActions.SET_BASE_CURRENCY_AMOUNT
                                    .type,
                                payload: values.baseCurrencyAmount,
                            });
                            handleConvertCurrency(values);
                        }}
                        validationSchema={Yup.object().shape({
                            baseCurrencyAmount: Yup.number()
                                .min(
                                    0.1,
                                    'Number must be greater than or equal to 0.1'
                                )
                                .required('This field is required'),
                            baseCurrency: Yup.object().required(
                                'This field is required'
                            ),
                            targetCurrency: Yup.object().required(
                                'This field is required'
                            ),
                        })}
                    >
                        <Form>
                            <StyledOuterGridContainer container>
                                <StyledInnerGridContainer container>
                                    <Grid container>
                                        <StyledGridItem item>
                                            <Field
                                                as={StyledInput}
                                                type="number"
                                                name="baseCurrencyAmount"
                                                placeholder="Type in amount of currency to convert from"
                                            />
                                            <ErrorMessage name="baseCurrencyAmount" />
                                        </StyledGridItem>
                                    </Grid>
                                    <Grid item>
                                        <Field name="baseCurrency">
                                            {({
                                                field,
                                                form,
                                            }: {
                                                field: any;
                                                form: any;
                                            }) => (
                                                <Autocomplete
                                                    {...field}
                                                    options={currencies}
                                                    getOptionLabel={(
                                                        option: Currency
                                                    ) => option.code || ''}
                                                    renderInput={(params) => (
                                                        <StyledTextField
                                                            {...params}
                                                            label="Choose Currency To Convert From"
                                                        />
                                                    )}
                                                    renderOption={(
                                                        props,
                                                        option: Currency
                                                    ) => (
                                                        <Box
                                                            component="li"
                                                            {...props}
                                                        >
                                                            {option.code}
                                                        </Box>
                                                    )}
                                                    onChange={(
                                                        event: React.ChangeEvent<{}>,
                                                        value: Currency | null
                                                    ) => {
                                                        if (value) {
                                                            form.setFieldValue(
                                                                'baseCurrency',
                                                                value
                                                            );
                                                        } else {
                                                            form.setFieldValue(
                                                                'baseCurrency',
                                                                ''
                                                            );
                                                        }
                                                    }}
                                                />
                                            )}
                                        </Field>
                                        <ErrorMessage name="baseCurrency" />
                                    </Grid>
                                    <Grid item>
                                        <Field name="targetCurrency">
                                            {({
                                                field,
                                                form,
                                            }: {
                                                field: any;
                                                form: any;
                                            }) => (
                                                <Autocomplete
                                                    {...field}
                                                    options={currencies}
                                                    getOptionLabel={(
                                                        option: Currency
                                                    ) => option.code || ''}
                                                    renderInput={(params) => (
                                                        <StyledTextField
                                                            {...params}
                                                            label="Choose Currency To Convert To"
                                                        />
                                                    )}
                                                    renderOption={(
                                                        props,
                                                        option: Currency
                                                    ) => (
                                                        <Box
                                                            component="li"
                                                            {...props}
                                                        >
                                                            {option.code}
                                                        </Box>
                                                    )}
                                                    onChange={(
                                                        event: React.ChangeEvent<{}>,
                                                        value: Currency | null
                                                    ) => {
                                                        if (value) {
                                                            form.setFieldValue(
                                                                'targetCurrency',
                                                                value
                                                            );
                                                        } else {
                                                            form.setFieldValue(
                                                                'targetCurrency',
                                                                ''
                                                            );
                                                        }
                                                    }}
                                                />
                                            )}
                                        </Field>
                                        <ErrorMessage name="targetCurrency" />
                                    </Grid>
                                </StyledInnerGridContainer>
                                <StyledIconButton type="submit">
                                    <StyledCurrencyExchangeIcon />
                                    <StyledTypographyButton variant="button">
                                        Convert
                                    </StyledTypographyButton>
                                </StyledIconButton>
                                <StyledInnerGridContainer container>
                                    <StyledGridItem item>
                                        <Field
                                            as={StyledInput}
                                            type="number"
                                            name="targetCurrencyAmount"
                                            placeholder="Result"
                                            value={targetCurrencyAmount}
                                        />
                                    </StyledGridItem>
                                </StyledInnerGridContainer>
                            </StyledOuterGridContainer>
                        </Form>
                    </Formik>
                </StyledCurrencyConverterBox>
            )}
        </Box>
    );
};

export default CurrencyConverter;
