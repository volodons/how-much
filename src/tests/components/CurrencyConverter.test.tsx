import {
    render,
    fireEvent,
    waitFor,
    screen,
    act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { currenciesActions } from '../../redux/ducks/currenciesDuck';
import { converterActions } from '../../redux/ducks/converterDuck';
import CurrencyConverter from '../../components/CurrencyConverter';

const createMockStore = configureMockStore();
const mockInitialState = {
    converter: {
        currencies: [],
        baseCurrency: { id: 1, code: 'UAH', exchangeRate: 1, featured: true },
        baseCurrencyAmount: 0,
        targetCurrency: '',
        targetCurrencyAmount: 0,
        exchangeRate: 0,
        error: null,
    },
    currencies: {
        currencies: [
            { id: 1, code: 'UAH', exchangeRate: 1, featured: true },
            { id: 2, code: 'USD', exchangeRate: 2, featured: false },
        ],
        currencyCodes: ['EUR', 'CAD'],
        baseCurrency: { id: 1, code: 'UAH', exchangeRate: 1, featured: true },
        error: null,
    },
};
const mockStore = createMockStore(mockInitialState);

describe('CurrencyConverter component', () => {
    it('should be a functional component', () => {
        expect(CurrencyConverter).toBeInstanceOf(Function);
    });

    it('should render Currency Converter heading', async () => {
        render(
            <Provider store={mockStore}>
                <CurrencyConverter />
            </Provider>
        );

        await waitFor(() => {
            expect(screen.getByText('Currency Converter')).toBeInTheDocument();
        });
    });

    it('should dispatch action to fetch currencies and base currency on mount', async () => {
        const mockDispatch = jest.fn();
        mockStore.dispatch = mockDispatch;

        render(
            <Provider store={mockStore}>
                <CurrencyConverter />
            </Provider>
        );

        await waitFor(() => {
            expect(mockDispatch).toHaveBeenCalledWith(
                expect.objectContaining({
                    type: converterActions.FETCH_CURRENCIES.type,
                })
            );
        });

        await waitFor(() => {
            expect(mockDispatch).toHaveBeenCalledWith(
                expect.objectContaining({
                    type: currenciesActions.FETCH_BASE_CURRENCY.type,
                })
            );
        });
    });

    it('should validate base currency amount field when submitting with an empty value', async () => {
        render(
            <Provider store={mockStore}>
                <CurrencyConverter />
            </Provider>
        );

        const convertButton = screen.getByText('Convert');

        act(() => {
            fireEvent.click(convertButton);
        });

        await waitFor(() => {
            expect(
                screen.getByText('This field is required')
            ).toBeInTheDocument();
        });
    });

    it('should update target currency amount field with the converted value', async () => {
        render(
            <Provider store={mockStore}>
                <CurrencyConverter />
            </Provider>
        );

        const baseCurrencyAmountInput = screen.getByPlaceholderText(
            'Type in amount of currency to convert from'
        );

        fireEvent.change(baseCurrencyAmountInput, { target: { value: '10' } });

        const convertButton = screen.getByText('Convert');

        act(() => {
            fireEvent.click(convertButton);
        });

        await waitFor(() => {
            expect(
                screen.getByPlaceholderText('Result').getAttribute('value')
            ).toBe('0');
        });
    });

    it('should match a snapshot', () => {
        const tree = renderer
            .create(
                <Provider store={mockStore}>
                    <CurrencyConverter />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
