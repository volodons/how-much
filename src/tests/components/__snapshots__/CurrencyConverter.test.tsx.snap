// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter component should match a snapshot 1`] = `
<div
  className="MuiBox-root css-0"
  data-testid="currency-converter-component"
>
  <div
    className="MuiBox-root css-c8c7jl"
  >
    <h2
      className="MuiTypography-root MuiTypography-h2 css-1oqg8so-MuiTypography-root"
    >
      Currency Converter
    </h2>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="MuiGrid-root MuiGrid-container css-k90tw8-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-17ifb59-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-item css-v78lst-MuiGrid-root"
            >
              <div
                className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1egh1qi-MuiInputBase-root-MuiInput-root"
                onClick={[Function]}
              >
                <input
                  className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  name="baseCurrencyAmount"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Type in amount of currency to convert from"
                  type="number"
                  value={0}
                />
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              aria-owns={null}
              className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
              name="baseCurrency"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
            >
              <div
                className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1mnimye-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink={false}
                  htmlFor=":r0:"
                  id=":r0:-label"
                >
                  Choose Currency To Convert From
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                  onClick={[Function]}
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-expanded={false}
                    aria-invalid={false}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    readOnly={false}
                    required={false}
                    role="combobox"
                    spellCheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Clear"
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onContextMenu={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Clear"
                      type="button"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="Open"
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onContextMenu={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </button>
                  </div>
                  <fieldset
                    aria-hidden={true}
                    className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="css-yjsfm1"
                    >
                      <span>
                        Choose Currency To Convert From
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              aria-owns={null}
              className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
              name="targetCurrency"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
            >
              <div
                className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1mnimye-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink={false}
                  htmlFor=":r2:"
                  id=":r2:-label"
                >
                  Choose Currency To Convert To
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                  onClick={[Function]}
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-expanded={false}
                    aria-invalid={false}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r2:"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    readOnly={false}
                    required={false}
                    role="combobox"
                    spellCheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Clear"
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onContextMenu={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Clear"
                      type="button"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="Open"
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onContextMenu={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </button>
                  </div>
                  <fieldset
                    aria-hidden={true}
                    className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="css-yjsfm1"
                    >
                      <span>
                        Choose Currency To Convert To
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-l401qg-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1iq9u08-MuiSvgIcon-root"
            data-testid="CurrencyExchangeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.89 11.1c-1.78-.59-2.64-.96-2.64-1.9 0-1.02 1.11-1.39 1.81-1.39 1.31 0 1.79.99 1.9 1.34l1.58-.67c-.15-.45-.82-1.92-2.54-2.24V5h-2v1.26c-2.48.56-2.49 2.86-2.49 2.96 0 2.27 2.25 2.91 3.35 3.31 1.58.56 2.28 1.07 2.28 2.03 0 1.13-1.05 1.61-1.98 1.61-1.82 0-2.34-1.87-2.4-2.09l-1.66.67c.63 2.19 2.28 2.78 2.9 2.96V19h2v-1.24c.4-.09 2.9-.59 2.9-3.22 0-1.39-.61-2.61-3.01-3.44M3 21H1v-6h6v2H4.52c1.61 2.41 4.36 4 7.48 4 4.97 0 9-4.03 9-9h2c0 6.08-4.92 11-11 11-3.72 0-7.01-1.85-9-4.67zm-2-9C1 5.92 5.92 1 12 1c3.72 0 7.01 1.85 9 4.67V3h2v6h-6V7h2.48C17.87 4.59 15.12 3 12 3c-4.97 0-9 4.03-9 9z"
            />
          </svg>
          <span
            className="MuiTypography-root MuiTypography-button css-xj5rr6-MuiTypography-root"
          >
            Convert
          </span>
        </button>
        <div
          className="MuiGrid-root MuiGrid-container css-17ifb59-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item css-v78lst-MuiGrid-root"
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1egh1qi-MuiInputBase-root-MuiInput-root"
              onClick={[Function]}
            >
              <input
                className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                name="targetCurrencyAmount"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Result"
                type="number"
                value={0}
              />
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
